### PR TITLE
http: auditing Path() calls for safety with Pathless CONNECT

### PR DESCRIPTION
**Pull request by alyssawilk**
This should result in all Path() calls not altered in #10720 being safe for path-less CONNECT.
The major change for this PR is that requests without a path will not be considered gRPC requests.  They're still currently rejected at the HCM, but when they are allowed through they will simply not be gRPC rather than causing crashes.

Risk Level: medium (L7 code refactor)
Testing: new unit tests
Docs Changes: n/a
Release Notes: n/a
Part of #1630 #1451

### CODE REVIEW

File: source/common/grpc/common.h [hunk 1]
```diff
@@ -36,6 +36,12 @@ class Common {
    */
   static bool hasGrpcContentType(const Http::RequestOrResponseHeaderMap& headers);
 
+  /**
+   * @param headers the headers to parse.
+   * @return bool indicating whether the header is a gRPC request header.
```
Comment (mattklein123) [inline source/common/grpc/common.h:41]: nit: what is a gRPC request header? Can you clarify? Maybe a more descriptive name also?

File: source/common/http/path_utility.cc [hunk 2]
```diff
@@ -29,6 +29,7 @@ absl::optional<std::string> canonicalizePath(absl::string_view original_path) {
 
 /* static */
 bool PathUtil::canonicalPath(RequestHeaderMap& headers) {
+  ASSERT(headers.Path());
```
Comment (yanavlasov) [inline source/common/http/path_utility.cc:32]: Is this invariant guaranteed?

File: source/extensions/tracers/xray/xray_tracer_impl.cc [hunk 3]
```diff
@@ -95,9 +95,11 @@ Tracing::SpanPtr Driver::startSpan(const Tracing::Config& config,
   }
 
   if (!should_trace.has_value()) {
+    absl::string_view path =
```
Comment (mattklein123) [inline source/extensions/tracers/xray/xray_tracer_impl.cc:98]: nit: const, similar elsewhere if possible.

File: source/common/grpc/common.h (hunk 1 above)
Comment (mattklein123) [reply source/common/grpc/common.h:41]: Sorry super pedantically can we call this `isGrpcRequestHeaders`? (It's confusing to understand if we are operating on a single header or a group of headers)

**Comment by alyssawilk** [comment]
Overnight fuzzing looks good!

**Pull request merged**